File: app/src/main/java/ru/stwtforever/fast/adapter/RecyclerAdapter.java
package ru.stwtforever.fast.adapter;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import ru.stwtforever.fast.util.ArrayUtil;

public abstract class RecyclerAdapter<T, VH extends RecyclerView.ViewHolder>
        extends RecyclerView.Adapter<VH> {
    private ArrayList<T> values;
    private ArrayList<T> cleanValues;

    protected Context context;
    LayoutInflater inflater;

    private OnItemClickListener click;
    private OnItemLongClickListener long_click;

    RecyclerAdapter(Context context, ArrayList<T> values) {
        this.context = context;
        this.values = values;

        this.inflater = LayoutInflater.from(context);
    }

    protected @ColorInt
    int getColor(int resId) {
        if (context == null) return -1;

        return context.getResources().getColor(resId);
    }

    @Nullable
    @Override
    public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return null;
    }

    @Override
    public void onBindViewHolder(@NonNull VH holder, int position) {
        updateListeners(holder.itemView, position);
    }

    @Override
    public int getItemCount() {
        return values == null ? 0 : values.size();
    }

    public void insert(ArrayList<T> items) {
        getValues().addAll(items);
    }

    public T getLastItem() {
        if (getItemCount() == 0) return getItem(0);
        return getItem(getItemCount() - 1);
    }

    public T getFirstItem() {
        return getItem(0);
    }

    public int getPosition(T item) {
        return getValues().indexOf(item);
    }

    public void changeItems(ArrayList<T> items) {
        this.values = new ArrayList<>(items);
    }

    public void add(int index, T item) {
        getValues().add(index, item);
    }

    public void add(T item) {
        getValues().add(item);
    }

    public void remove(int position) {
        getValues().remove(position);
    }

    public void clear() {
        if (!ArrayUtil.isEmpty(getValues()))
            getValues().clear();
    }

    public void setOnItemClickListener(OnItemClickListener l) {
        this.click = l;
    }

    public void setOnItemLongClickListener(OnItemLongClickListener l) {
        this.long_click = l;
    }

    private void updateListeners(View v, final int position) {
        if (click != null) {
            v.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    click.onItemClick(v, position);
                }
            });
        }

        if (long_click != null) {
            v.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    long_click.onItemLongClick(v, position);
                    return click != null;
                }
            });
        }
    }

    public T getItem(int position) {
        if (position == getItemCount())
            position--;
        return values.get(position);
    }

    public void filter(String query) {
        String lowerQuery = query.toLowerCase();

        if (cleanValues == null) {
            cleanValues = new ArrayList<>(values);
        }
        values.clear();

        if (query.isEmpty()) {
            values.addAll(cleanValues);
        } else {
            for (T value : cleanValues) {
                if (onQueryItem(value, lowerQuery)) {
                    values.add(value);
                }
            }
        }

        notifyItemRangeChanged(0, getItemCount());
    }

    protected boolean onQueryItem(T item, String lowerQuery) {
        return false;
    }

    public ArrayList<T> getValues() {
        return values;
    }

    public String getString(int res) {
        return context.getString(res);
    }

    public String getString(int res, Object... args) {
        return context.getString(res, args);
    }

    public Drawable getDrawable(int res) {
        return context.getDrawable(res);
    }

    public boolean isEmpty() {
        return getItemCount() == 0;
    }

    public interface OnItemClickListener {
        void onItemClick(View v, int position);
    }

    public interface OnItemLongClickListener {
        void onItemLongClick(View v, int position);
    }

}

File: app/src/main/java/ru/melodin/fast/adapter/RecyclerAdapter.kt
package ru.melodin.fast.adapter

import android.content.Context
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.ColorInt
import androidx.collection.SparseArrayCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import ru.melodin.fast.model.Model


abstract class RecyclerAdapter<T : Model, VH : RecyclerHolder> internal constructor(protected var context: Context, var viewRes: Int, values: ArrayList<T>) : RecyclerView.Adapter<RecyclerHolder>() {


    var values: ArrayList<T>? = null
        private set

    private var inflater: LayoutInflater
    private var cleanValues: ArrayList<T>? = null
    private var click: OnItemClickListener? = null
    private var longClick: OnItemLongClickListener? = null

    private val selectedItems = SparseArrayCompat<Model>()

    val isSelected: Boolean
        get() = selectedCount > 0

    val selectedMessages: ArrayList<T>
        get() {
            val items = ArrayList<T>()

            for (item in values!!)
                if (item.isSelected)
                    items.add(item)

            return items
        }

    open val selectedCount: Int
        get() = selectedItems.size()

    val isEmpty: Boolean
        get() = values == null || values!!.size == 0

    val lastPosition: Int
        get() = values!!.size - 1

    init {
        this.values = values

        this.inflater = LayoutInflater.from(context)
    }

    fun setSelected(position: Int, selected: Boolean) {
        val item = getItem(position)
        item.isSelected = selected

        if (selected) {
            selectedItems.append(position, item)
        } else {
            selectedItems.remove(position)
        }

        notifyItemChanged(position, -1)
    }


    fun clearSelected() {
        for (item in values!!) {
            item.isSelected = false
        }

        selectedItems.clear()
    }

    open fun toggleSelected(position: Int) {
        val item = getItem(position)

        val selected = !item.isSelected
        item.isSelected = selected

        if (selected) {
            selectedItems.append(position, item)
        } else {
            selectedItems.remove(position)
        }
    }


    @ColorInt
    protected fun getColor(resId: Int): Int {
        return ContextCompat.getColor(context, resId)
    }

    override fun onBindViewHolder(holder: RecyclerHolder, position: Int) {
        updateListeners(holder.itemView, position)
        holder.bind(position)
    }

    fun getView(parent: ViewGroup?): View? {
        return inflater.inflate(viewRes, parent, false)
    }

    override fun getItemCount(): Int {
        return if (values == null) 0 else values!!.size
    }

    fun insert(items: ArrayList<T>) {
        values!!.addAll(items)
    }

    open fun changeItems(items: ArrayList<T>) {
        this.values!!.clear()
        notifyDataSetChanged()
        this.values = ArrayList(items)
    }

    fun add(index: Int, item: T) {
        values!!.add(index, item)
    }

    fun add(item: T) {
        values!!.add(item)
    }

    fun remove(position: Int) {
        values!!.removeAt(position)
    }

    fun clear() {
        values!!.clear()
    }

    fun setOnItemClickListener(l: OnItemClickListener) {
        this.click = l
    }

    fun setOnItemLongClickListener(l: OnItemLongClickListener) {
        this.longClick = l
    }

    private fun updateListeners(v: View, position: Int) {
        if (click != null) {
            v.setOnClickListener { click!!.onItemClick(position) }
        }

        if (longClick != null) {
            v.setOnLongClickListener {
                longClick!!.onItemLongClick(position)
                click != null
            }
        }
    }

    open fun getItem(position: Int): T {
        return values!![position]
    }

    fun getString(res: Int): String {
        return context.getString(res)
    }

    fun getString(res: Int, vararg args: Any): String {
        return context.getString(res, *args)
    }

    protected fun getDrawable(res: Int): Drawable? {
        return ContextCompat.getDrawable(context, res)
    }

    open fun isSelected(position: Int): Boolean {
        return getItem(position).isSelected
    }

    interface OnItemClickListener {
        fun onItemClick(position: Int)
    }

    interface OnItemLongClickListener {
        fun onItemLongClick(position: Int)
    }

}